homeassistant/components/generic/diagnostics.py

```diff
@@ -0,0 +1,41 @@
+"""Diagnostics support for generic (IP camera)."""
+from __future__ import annotations
+
+import re
+from typing import Any
+
+from homeassistant.components.diagnostics import async_redact_data
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+
+from .const import CONF_STILL_IMAGE_URL, CONF_STREAM_SOURCE
+
+TO_REDACT = {
+    CONF_PASSWORD,
+    CONF_USERNAME,
+}
+
+# A very similar redact function is in components.sql.  Possible to be made common.
+URL_RE = re.compile("//.*:.*@")
+
+
+def redact_credentials(data: str) -> str:
+    """Redact credentials from string data."""
+    return URL_RE.sub("//****:****@", data)
+
```

Comment:
It would be cleaner to use YARL to remove the creds.  I realize its a bit slower, but security is usually worth trading performance

Reply:
I agree it should be a generic helper